File: Conceitos de LPs/Subprogramas/Matrizes como Parametros/Matrizes_como_parametros.cpp
//C++
//Parametro Real: m
//Parametro Formal: matriz
//Passagem de Parâmetro do tipo Matriz é, por padrão, por Referência

#include <iostream>
using namespace std;

// Função para imprimir a matriz
void imprimirMatriz(int matriz[2][4], int linhas, int colunas) {
    for (int i = 0; i < linhas; i++) {
        for (int j = 0; j < colunas; j++) {
            cout << matriz[i][j] << " ";
        }
        cout << endl;
    }
}

// Função para dobrar os elementos da matriz
void dobrarMatriz(int matriz[2][4], int linhas, int colunas) {
    for (int i = 0; i < linhas; i++) {
        for (int j = 0; j < colunas; j++) {
            matriz[i][j] *= 2;
        }
    }
}

int main() {
    int m[2][4] = {{1, 2, 3, 4}, {5, 6, 7, 8}};
    
    cout << "Antes de dobrar:" << endl;
    imprimirMatriz(m, 2, 4);

    dobrarMatriz(m, 2, 4);

    cout << "Depois de dobrar:" << endl;
    imprimirMatriz(m, 2, 4);
    
    return 0;
}

File: Conceitos de LPs/Subprogramas/Matrizes como Parametros/MatrizesComoParametros.kt
//Kotlin
//Parametro Real: m
//Parametro Formal: matriz
//Passagem de Parâmetro do tipo Matriz é, por padrão, por Referência

// Função para imprimir a matriz
fun imprimirMatriz(matriz: Array<IntArray>) {
    for (linha in matriz) {
        println(linha.joinToString(" "))
    }
}

// Função para dobrar os elementos da matriz
fun dobrarMatriz(matriz: Array<IntArray>) {
    for (i in matriz.indices) {
        for (j in matriz[i].indices) {
            matriz[i][j] *= 2
        }
    }
}

fun main() {
    val m = arrayOf(intArrayOf(1, 2, 3, 4), intArrayOf(5, 6, 7, 8))
    
    println("Antes de dobrar:")
    imprimirMatriz(m)

    dobrarMatriz(m)

    println("Depois de dobrar:")
    imprimirMatriz(m)
}
